rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,238 @@
+
+The DNS policy will have a finalizer, when it is deleted the controller will mark the created DNS Records for deletion.
+
+The DNS Records will have a finalizer from the DNS Operator which will clean the records from the zone before removing the finalizer.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## General Logic Overview
+
+The general flow for in the Kuadrant operator follows a single path, where it will always output a DNS Record which specifies everything needed for this workload on this cluster, and is unaware of the concept of distributed DNS, or phrased more simply:
+Build a DNS Record (let's refer to this as a kuadrantRecord) based on the local gateways and DNS Policy.
+
+This kuadrantRecord is reconciled by the DNS Operator, where there are 2 possible routes:
+
+Where a multiClusterID is specified in the relevant ProviderSecret the DNS Operator will act with the DNS Provider's zone as the authority and ensure that the records in the zone relevant to the hosts defined within gateways on it's cluster are accurate. When cleaning up a DNS Record in this scenario the operator will ensure all records owned by the policy for the gateway on the local cluster are removed and then prune unresolvable records related to the hosts defined in the gateway from the DNS Provider zone.
+
+Where a multiClusterID is not specified the DNS Operator will act with the DNS Record as the authority and ensure the entire host is updated in the zone to exactly match the specified DNS Record when cleaning up in this scenario, the host is fully removed from the DNS Provider zone.
+
+### API Updates
+New configuration options in the providerSecret:
+- `multiClusterID`
+  - When absent the DNS Operator will consider itself the owner of the zone and the local DNS Records as the authority.
+  - When present the DNS Operator will consider the zone the authority and only insert/update any missing records from the local DNS Records.
+  - This value will be hashed whenever it is used in the zone file i.e. in the A/CNAME records and also in the health check TXT record)
+  - Used when generating an A record, e.g. `<hash>-<multiClusterID>.<lb-id>.<host...>`
+- `RefreshInterval`
+  - When absent it defaults to the shortest TTL of created records. This is how often in seconds to reconcile the DNS Policy in order to pull fresh zone values from the DNS Provider.
+
+New fields added to the DNS Policy CRD: 
+- `HealthCheck.MultiClusterFailureThreshold`
+  - When absent defaults to `100`, what percentage of clusters need to be reporting an endpoint unhealthy on a remote IP for it to be removed from the zone by a cluster that does not own that IP.
+
+New fields added to the DNS Record CRD:
+- `HealthCheck`
+  - Propagated from the DNS Policy, to facilitate reconciling health checks in the DNS controller.
+- `ProviderSecret`
+  - Propagated from the DNS Policy to facilitate pulling the zone in the DNS controller.
```

Comment:
This isn't new, and already part of this RFC  https://github.com/Kuadrant/architecture/blob/main/rfcs/0005-single-cluster-dnspolicy.md#dnsrecordspecproviderref

Reply:
ah, I thought it might be, good. I'll remove it from here.